pytensor/link/numba/dispatch/basic.py

```diff
@@ -593,17 +594,49 @@ def inputs_cast(x):
 
 @numba_funcify.register(Dot)
 def numba_funcify_Dot(op, node, **kwargs):
-    # Numba's `np.dot` does not support integer dtypes, so we need to cast to
-    # float.
+    # Numba's `np.dot` does not support integer dtypes, so we need to cast to float.
+    x, y = node.inputs
+    [out] = node.outputs
 
-    out_dtype = node.outputs[0].type.numpy_dtype
-    inputs_cast = int_to_float_fn(node.inputs, out_dtype)
+    x_dtype = x.type.dtype
+    y_dtype = y.type.dtype
+    dot_dtype = f"float{max((32, out.type.numpy_dtype.itemsize * 8))}"
+    out_dtype = out.type.dtype
 
-    @numba_njit
-    def dot(x, y):
-        return np.asarray(np.dot(inputs_cast(x), inputs_cast(y))).astype(out_dtype)
+    if x_dtype == dot_dtype and y_dtype == dot_dtype:
```

Comment:
I need these branches otherwise, I get some `failed to unify` during numba compilation. 
It doesn't like the pattern:
```python
if x.dtype != dot_dtype:
  x = x.astype(dot_dtype)
```
When it's actually needed. Can be simplified once `astype(copy=False)` is implemented in numba